### ISSUE TITLE
"On this page" navigation behaves differently in English and French

### ISSUE THREAD

**Issue by tristan-orourke** [GCTC-NTGC/gc-digital-talent#4378]
On the English version of the Profile page, clicking a link in the On this Page section smoothly scrolls to the correct section.
https://recordit.co/NOn9oNntoG

On the French version of the page, clicking an On this Page link causes a full page refresh, and leaves you in slightly the wrong position!
https://recordit.co/dvQc9bKEKv

**Comment by esizer** [comment]
Has something to do with `navigate` when on FR. If you comment out the following line, it works fine but does not add the `#id` to the pathname.

https://github.com/GCTC-NTGC/gc-digital-talent/blob/e9e70dded6ff0bd1ed0d53703aadd5dc208b757f/frontend/common/src/components/Link/ScrollToLink.tsx#L22

**Comment by esizer** [comment]
Okay, took a bit of time and looked into this more. It seems this is coming from old faithful `useLocation`.

<img width="765" alt="Screen Shot 2022-10-21 at 2 53 37 PM" src="https://user-images.githubusercontent.com/4127998/197269163-4d2920dd-fc8e-421f-a4d9-04aaf3c22ae6.png">

The page is not reloading at all, it is re-rendering based on changed state in our LanguageRedirectContext. The profiler is showing a re-render because of a hook change. The only hooks that appear in that component are `useLocation` and `useIntlLanguages`.

https://github.com/GCTC-NTGC/gc-digital-talent/blob/1484e37a428c51fd6764e744ecbcca543cf6e0f5/frontend/common/src/components/LanguageRedirectContainer/LanguageRedirectContainer.tsx#L38-L43

## What I believe is happening?

1. Location changes (`locaton.hash`)
2. New location is set
3. `LanguageRedirectContainer` sees new location and triggers a re-render

## Why isn't this happening on the English version?

Good question! Maybe due to use defaulting to English and less logic to determine the locale since it is assumed by default.? 🤷‍♂️ 

## Possible Solution

So, i think we can solve this by comparing the locations in `useLocation` a bit better to reduce the location updating. I think the only thing we care about is `pathname` and `state`. Though, I could be wrong.

https://github.com/GCTC-NTGC/gc-digital-talent/blob/1484e37a428c51fd6764e744ecbcca543cf6e0f5/frontend/common/src/helpers/router.tsx#L23-L25

The previous code could be changed to the following, to reduce re-renders.

```tsx
if(location.pathname !== newLocation.pathname && location.state !== newLocation.state) {
  setLocation(newLocation)
}
```

I tested it and this does work but I'm not sure if it interferes with other functionality.

## Another (possible) solution

Prioritize the `react-router` ticket since they include almost all of our `router.tsx` logic and has likely handled this use case...maybe.

**Comment by tristan-orourke** [comment]
> I think we can solve this by comparing the locations in useLocation a bit better to reduce the location updating. I think the only thing we care about is `pathname` and `state`.

I'd rather not modify useLocation like that. Even though we're usually only using pathname, we might use other fields in other cases. Instead, can we let React know that LanguageRedirectContainer only cares about those two fields? Will replacing
```
 const location = useLocation(); 
 ```
 with
 ```
 const {pathname, state} = useLocation();
```
solve the extra re-renders?

**Comment by esizer** [comment]
> Instead, can we let React know that LanguageRedirectContainer only cares about those two fields? Will replacing

Unfortunately not because we need the entire location in that component for `localizePath`:

https://github.com/GCTC-NTGC/gc-digital-talent/blob/295d4d7465bcc479973db27e9a650736300cf037/frontend/common/src/components/LanguageRedirectContainer/LanguageRedirectContainer.tsx#L51

https://github.com/GCTC-NTGC/gc-digital-talent/blob/295d4d7465bcc479973db27e9a650736300cf037/frontend/common/src/helpers/localize.ts#L23-L49

**Comment by esizer** [comment]
Closed by #4597
